R/card.R

```diff
@@ -0,0 +1,298 @@
+        } else {
+          these_children
+        }
+      }, SIMPLIFY = FALSE)
+    children <- unlist(children, recursive = FALSE)
+  }
+
+  tag <- htmltools::tags$div(
+    class = "card",
+    class = class,
+    style = htmltools::css(width = width, height = height),
+    !!!attribs,
+    !!!children
+  )
+  as_fragment(
+    tag_require(tag, version = 4, caller = "card()")
+  )
+}
+
+#' @rdname card_body
+#' @export
+as.card_item <- function(x) {
+  class(x) <- c("card_item", class(x))
+  x
+}
+
+#' @rdname card_body
+#' @export
+is.card_item <- function(x) {
+  inherits(x, "card_item")
+}
+
+#' Card components
+#'
+#' @description This topic describes various components that are intended to go
+#'   directly inside of a [card()]. These components can be used in combination;
+#'   for example, a single card could contain a `card_header()`, multiple
+#'   `card_body()`s, a `card_list()`, and finally a `card_footer()`.
+#'
+#' @param ... Named arguments become attributes on the `<div class="card">`
+#'   element. Unnamed arguments become card items, and can be any valid child of
+#'   an [htmltools tag][htmltools::tags].
+#' @param class Additional CSS classes to include on the card div.
+#' @param padding Whether vertical and/or horizontal padding should be included.
+#'   Use `NULL` for neither, `"x"` for horizontal only, `"y"` for vertical only,
+#'   or `c("x", "y")` for both (the default).
+#' @param stretch Set to `TRUE` if this `card_body` is eager to use any extra
+#'   vertical space is available in the card.
+#'
+#' @export
+card_body <- function(..., class = NULL, padding = c("x", "y"), stretch = FALSE) {
+  res <- htmltools::tags$div(class = "card-body",
+    class = if (!"x" %in% padding) "px-0",
+    class = if (!"y" %in% padding) "py-0",
+    class = class,
+    ...,
+    style = if (!stretch) css(
+      flex = "0",
+      `_webkit_flex` = "0"
+    ),
+  )
+  as.card_item(res)
+}
+
+#' @export
+card_body_scroll <- function(..., class = NULL, height = NULL, padding = c("x", "y")) {
+  res <- htmltools::tags$div(class = "card-body card-body-scroll",
+    class = if (!"x" %in% padding) "px-0",
+    class = if (!"y" %in% padding) "py-0",
+    class = class,
+    style = css(
+      # May be NULL
+      flex_basis = validateCssUnit(height),
+      `-webkit-flex-basis` = validateCssUnit(height)
+    ),
+    ...
+  )
+  as.card_item(res)
+}
+
+#' @rdname card_body
+#' @export
+card_header <- function(..., class = NULL) {
+  res <- htmltools::tags$div(class = "card-header", class = class,
+    ...
+  )
+  as.card_item(res)
+}
+
+#' @rdname card_body
+#' @export
+card_heading <- function(..., class = NULL, container = htmltools::tags$h6) {
+  res <- container(class = "card-header mt-0", class = class,
+    ...
+  )
+  as.card_item(res)
+}
+
+#' @rdname card_body
+#' @export
+card_title <- function(..., class = NULL) {
+  res <- htmltools::tags$h5(class = "card-header", class = class,
+    ...
+  )
+  as.card_item(res)
+}
```

Comment:
Did you intend to have 3 versions of essentially the same thing? I think I would be in favor of just having  (especially since you can easily control the size with the `.display-*` utility classes)

```suggestion
#' @rdname card_body
#' @export
card_header <- function(..., class = NULL, container = htmltools::div) {
  res <- container(class = "card-header", class = class,
    ...
  )
  as.card_item(res)
}
```

Reply:
I think I intended to have two (one generic `card-header` and one that takes a string and puts it in a heading) but not two. Certainly `card_heading` and `card_title` seem redundant. I'm open to whatever though.